interpreter.go

```diff
@@ -645,26 +612,112 @@ func (i *interpreter) manifestJSON(loc *LocationRange, v value, multiline bool,
 		buf.WriteString(unparseString(v.value))
 
 	default:
-		return makeRuntimeError(loc, fmt.Sprintf("Manifesting this value not implemented yet: %s", reflect.TypeOf(v)))
+		return makeRuntimeError(
+			fmt.Sprintf("Manifesting this value not implemented yet: %s", reflect.TypeOf(v)),
+			i.getCurrentStackTrace(trace),
+		)
 
 	}
 	return nil
 }
 
-func evaluate(ast astNode, ext vmExtMap, maxStack int) (string, error) {
+func (i *interpreter) EvalInCleanEnv(fromWhere *TraceElement, newContext *TraceContext,
+	env *environment, ast astNode) (value, error) {
+	err := i.newCall(fromWhere, *env)
+	if err != nil {
+		return nil, err
+	}
+	val, err := i.evaluate(ast, newContext)
+	i.stack.pop()
+	return val, err
+}
+
+func buildStdObject(i *interpreter) (value, error) {
+	objVal, err := evaluateStd(i)
+	if err != nil {
+		return nil, err
+	}
+	obj := objVal.(*valueSimpleObject)
+	builtinFields := map[string]unboundField{
+		// TODO(sbarzowski) not a real thing, just the simples thing I could put there
+		"nullValue": &readyValue{&valueNull{}},
+	}
+	for key, ec := range funcBuiltins {
+		function := valueFunction{ec: ec} // TODO(sbarzowski) better way to build function value
+		builtinFields[key] = &readyValue{&function}
+	}
+
+	for name, value := range builtinFields {
+		obj.fields[name] = valueSimpleObjectField{astObjectFieldHidden, value}
+	}
+	return obj, nil
+}
+
+func evaluateStd(i *interpreter) (value, error) {
+	beforeStdEnv := makeEnvironment(
+		bindingFrame{},
+		selfBinding{
+			nil,
+			123456789, // poison value
+		},
+	)
+	evalLoc := makeLocationRangeMessage("During evaluation of std")
+	evalTrace := &TraceElement{loc: &evalLoc}
+	ast, err := snippetToAST("std.jsonnet", getStdCode())
+	if err != nil {
+		return nil, err
+	}
+	context := TraceContext{Name: "<stdlib>"}
+	return i.EvalInCleanEnv(evalTrace, &context, &beforeStdEnv, ast)
+}
+
+func buildInterpreter(ext vmExtMap, maxStack int, importer Importer) (*interpreter, error) {
 	i := interpreter{
 		stack:          makeCallStack(maxStack),
 		idArrayElement: identifier("array_element"),
 		idInvariant:    identifier("object_assert"),
 		externalVars:   ext,
+
+		importCache: MakeImportCache(importer),
+	}
+
+	stdObj, err := buildStdObject(&i)
+	if err != nil {
+		return nil, err
 	}
-	result, err := i.evaluate(ast)
+
+	i.initialEnv = makeEnvironment(
```

Comment:
The empty selfBinding with poison value comes up twice.  Maybe refactor it to makeUnboundSelfBinding or even makeEmptyEnvironment (in which case you would assign the "std" field after it in this case)

Reply:
Done.

BTW Currently we have 3 places in which we do a very similar thing - import, std evaluation and main evaluation. I may try to do something about it later. Not sure if it won't just complicate things...